File: app/src/main/java/org/dhis2/utils/ValidationUtils.java
package org.dhis2.utils;

import org.hisp.dhis.android.core.common.ValueType;

import java.util.Locale;

public class ValidationUtils {

    public static String validate(ValueType valueType, String value) {
        if (value == null)
            return null;

        String validatedValue = value;

        switch (valueType) {
            case INTEGER:
            case INTEGER_NEGATIVE:
            case INTEGER_POSITIVE:
            case INTEGER_ZERO_OR_POSITIVE:
            case PERCENTAGE:
                validatedValue = String.format(Locale.US, "%.0f", Float.valueOf(value));
                break;
            case UNIT_INTERVAL:
                validatedValue = Float.valueOf(value).toString();
                break;
            case NUMBER:
                validatedValue = String.format(Locale.US, "%.1f", Double.valueOf(value));
                break;
        }

        return validatedValue;
    }
}

File: app/src/main/java/org/dhis2/utils/ValidationUtils.kt
package org.dhis2.utils

import org.hisp.dhis.android.core.common.ValueType

class ValidationUtils {

    companion object {

        @JvmStatic
        fun validate(valueType: ValueType, value: String?): String? {
            return value?.let {
                if (value == "-") {
                    "0"
                } else {
                    when (valueType) {
                        ValueType.INTEGER,
                        ValueType.INTEGER_NEGATIVE,
                        ValueType.INTEGER_POSITIVE,
                        ValueType.INTEGER_ZERO_OR_POSITIVE,
                        ValueType.PERCENTAGE,
                        -> it.toIntOrNull()?.toString() ?: value
                        ValueType.UNIT_INTERVAL -> it.toFloatOrNull()?.toString() ?: value
                        ValueType.NUMBER ->
                            if (it.toDouble() % 1.toDouble() == 0.toDouble()) {
                                it.toIntOrNull()?.toString() ?: value
                            } else {
                                it.toDouble().toString()
                            }
                        else -> it
                    }
                }
            }
        }
    }
}
